theano/tensor/subtensor.py

```diff
@@ -1934,7 +1934,7 @@ def copy_of_x(self, x):
         # max_depth: we pass 0 to have this parameter ignored
         # requirements: here we pass NPY_ARRAY_ENSURECOPY to force a copy
         # context: this is almost always NULL, I'm not sure what it's used for
-        return """(PyArrayObject*)PyArray_FromAny(py_%(x)s, NULL, 0, 0,
+        return """(PyArrayObject*)PyArray_FromAny(py_%(x)s, PyArray_DTYPE(%(x)s), 0, 0,
```

Comment:
According to the doc, `PyArray_DTYPE` returns a borrowed reference, and `PyArray_FromAny` steals it, so we need to call `Py_INCREF` on that dtype.